assets/js/linear_algebra/dot_product_project2d.js

```diff
@@ -0,0 +1,198 @@
+var dot_product_project2d = (function() {
+
+
+
+var origin = [150, 140], 
+  scale = 60, 
+  scatter = [], 
+  axis = [],
+  expectedAxis = [],
+  startAngleX = Math.PI,
+  startAngleY = 0.,
+  startAngleZ = 0.
+  axis_len = 2;
+
+let unit = axis_len/10;
+
+var svg = d3.select("#svg_dot_product_project2d");
+
+
+var lib = space_plot_lib(
+  svg,
+  origin, 
+  scale,
+  is_2d=true);
+
+
+svg = svg.call(d3.drag()
+         .on('drag', dragged)
+         .on('start', drag_start)
+         .on('end', drag_end))
+         .append('g');
+
+axis = lib.init_float_axis(axis_len=axis_len, unit=unit);
+
+
+function plot(scatter, axis, tt){
+
+  basis = {
+      ex: axis[4][1], 
+      ey: axis[14][1],
+      // ex: axis[axis_len/unit*4 * 0][1], 
+      // ey: axis[axis_len/unit*14 * 1][1],
+      ez: 0.
+  };
+
+  scatter.forEach(function(d, i){
+    var coord = lib.dot_basis(d, basis);
+    if (i == 0) {
+      d.text = 'u = ';
+    } else {
+      d.text = 'v = ';
+    }
+    d.text += '['.concat(
+      coord.x.toFixed(2),
+      ', ',
+      coord.y.toFixed(2),
+      ']');
+    
+    d.text_color = d.color;
+
+  })
+
+  lib.plot_lines(axis);
+
+  var lines = [];
+  scatter.forEach(function(d){
+    lines.push([
+        {x: 0., y: 0., z: 0.},
+        {x: d.x, y: d.y, z: d.z, 
+         color: d.color, tt: true}
+    ]);
+  });
+
+  let mult_vec = scatter[0].x * scatter[1].x +
+                 scatter[0].y * scatter[1].y;
+  let projectionx = scatter[1].x * mult_vec;
+  let projectiony = scatter[1].y * mult_vec;
+
+  lines.push([
```

Comment:
please create a new variable "projection_line" instead of referring to it as lines[2] to avoid future bugs.

Reply:
done